core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java

```diff
@@ -0,0 +1,471 @@
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+/**
+ * This class is a REST handler for Amazon S3 API.
+ */
+@NotThreadSafe
+@Path(S3RestServiceHandler.SERVICE_PREFIX)
+@Produces(MediaType.APPLICATION_XML)
+@Consumes(MediaType.APPLICATION_XML)
+public final class S3RestServiceHandler {
+  public static final String SERVICE_PREFIX = "s3";
+
+  /**
+   * Bucket must be a directory directly under a mount point. If it is under a non-root mount point,
+   * the bucket separator must be used as the separator in the bucket name, for example,
+   * mount:point:bucket represents Alluxio directory /mount/point/bucket.
+   */
+  public static final String BUCKET_SEPARATOR = ":";
+
+  /* Bucket is the first component in the URL path. */
+  public static final String BUCKET_PARAM = "{bucket}/";
+  /* Object is after bucket in the URL path */
+  public static final String OBJECT_PARAM = "{bucket}/{object:.+}";
+
+  private final FileSystem mFileSystem;
+
+  /**
+   * Constructs a new {@link S3RestServiceHandler}.
+   *
+   * @param context context for the servlet
+   */
+  public S3RestServiceHandler(@Context ServletContext context) {
+    mFileSystem =
+        (FileSystem) context.getAttribute(ProxyWebServer.FILE_SYSTEM_SERVLET_RESOURCE_KEY);
+  }
+
+  /**
+   * @summary creates a bucket
+   * @param bucket the bucket name
+   * @return the response object
+   */
+  @PUT
+  @Path(BUCKET_PARAM)
+  @ReturnType("java.lang.Void")
+  public Response createBucket(@PathParam("bucket") final String bucket) {
+    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Response.Status>() {
+      @Override
+      public Response.Status call() throws S3Exception {
+        Preconditions.checkNotNull(bucket, "required 'bucket' parameter is missing");
+        String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
+
+        // Create the bucket.
+        CreateDirectoryOptions options = CreateDirectoryOptions.defaults()
+            .setWriteType(getS3WriteType());
+        try {
+          mFileSystem.createDirectory(new AlluxioURI(bucketPath), options);
+        } catch (Exception e) {
+          throw toBucketS3Exception(e, bucketPath);
+        }
+        return Response.Status.OK;
+      }
+    });
+  }
+
+  /**
+   * @summary deletes a bucket
+   * @param bucket the bucket name
+   * @return the response object
+   */
+  @DELETE
+  @Path(BUCKET_PARAM)
+  @ReturnType("java.lang.Void")
+  public Response deleteBucket(@PathParam("bucket") final String bucket) {
+    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Response.Status>() {
+      @Override
+      public Response.Status call() throws S3Exception {
+        Preconditions.checkNotNull(bucket, "required 'bucket' parameter is missing");
+        String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
+
+        checkBucketIsAlluxioDirectory(bucketPath);
```

Comment:
if check bucket throws an exception do we return the correct error response?

Reply:
yes, it's done in line 427 `throw toBucketS3Exception(e, bucketPath);` and then  `S3RestUtils#call` handles `S3Exception`